crates/constraint-vm/src/access.rs

```diff
@@ -286,42 +275,69 @@ pub(crate) fn repeat_counter(stack: &mut Stack, repeat: &Repeat) -> OpResult<()>
     Ok(stack.push(counter)?)
 }
 
-pub(crate) fn transient(stack: &mut Stack, solution: SolutionAccess) -> OpResult<()> {
-    let pathway = stack.pop()?;
-    let pathway =
-        SolutionDataIndex::try_from(pathway).map_err(|_| AccessError::TransientDataOutOfBounds)?;
-    let value = stack.pop_len_words::<_, _, StackError>(|key| {
-        let value = solution
-            .transient_data
-            .get(&pathway)
-            .ok_or(StackError::IndexOutOfBounds)?
-            .get(key)
-            .ok_or(StackError::IndexOutOfBounds)?;
-        Ok(value.clone())
-    })?;
+/// `Access::PubVar` implementation.
+pub(crate) fn pub_var(stack: &mut Stack, pub_vars: &TransientData) -> OpResult<()> {
+    // Pop the value_len, value_ix and create a range.
+    let value_len = stack
+        .pop()
+        .map_err(|_| MissingAccessArgError::PubVarValueLen)?;
+    let value_ix = stack
+        .pop()
+        .map_err(|_| MissingAccessArgError::PubVarValueIx)?;
+    let range = range_from_start_len(value_ix, value_len).ok_or(AccessError::InvalidAccessRange)?;
+
+    // Pop the key and access the value.
+    let value = stack
+        .pop_len_words_with_additional::<_, _, OpError>(1, |slice| {
+            let (pathway_ix, key) = slice
+                .split_first()
+                .expect("Can't fail because must have at least 1 word");
+            let pathway_ix = SolutionDataIndex::try_from(*pathway_ix)
+                .map_err(|_| AccessError::PathwayOutOfBounds(*pathway_ix))?;
+            let value = pub_vars
+                .get(&pathway_ix)
+                .ok_or(AccessError::PathwayOutOfBounds(pathway_ix as Word))?
+                .get(key)
+                .ok_or(AccessError::PubVarKeyOutOfBounds)?
+                .get(range)
+                .ok_or(AccessError::PubVarDataOutOfBounds)?;
+            Ok(value.to_vec())
+        })
+        .map_err(map_key_len_err)?;
```

Comment:
Nice.

An alternative to `pop_len_words_with_additional` that might be a bit simpler / generally useful could be something like:

```rust
    /// Pop `n` words from the stack and provide them to the given function.
    pub fn pop_words<F, O, E>(&mut self, n: usize, f: F) -> Result<O, E>
    where
        F: FnOnce(&[Word]) -> Result<O, E>,
        E: From<StackError>,
    {
        let len = self.len();
        let ix = len.checked_sub(n).ok_or(StackError::IndexOutOfBounds)?;
        let (rest, slice) = self.split_at(ix);
        let output = f(slice)?;
        self.0.truncate(rest.len());
        Ok(output)
    }
```

Then we could just pop the key length and call `pop_words(key_len.checked_add(1)?, |slice| ...)`